fms/models/mixtral.py

```diff
@@ -0,0 +1,398 @@
+        distributed_strategy: DistributedStrategy = NoOpStrategy,
+        **kwargs,
+    ):
+        super(Mixtral, self).__init__()
+        if config is not None:
+            self.config = config
+        else:
+            self.config = MixtralConfig()
+        self.config = self.config.updated(**kwargs)
+        self.distributed_strategy = distributed_strategy
+
+        self.width = self.config.dim
+        self.pad_id = self.config.pad_id
+        self.max_expected_seq_len = self.config.max_expected_seq_len
+
+        shared = WordEmbedding(
+            self.config.src_vocab_size,
+            self.config.dim,
+            padding_idx=self.config.pad_id,
+            abs_pos=False,
+            reversible=True,
+            tie_weights=False,
+            bias=False,
+        )
+        self.shared = self.distributed_strategy.distribute_module(shared)
+
+        self.rot_emb = RotaryEmbedding(
+            dim=self.config.dim // self.config.nheads,
+            ratio=self.config.rope_base,
+            ntk_scaling=self.config.ntk_scaling,
+            max_seq_len=self.config.max_expected_seq_len,
+        )
+        if isinstance(self.distributed_strategy, UniformModelParallelStrategy):
+            for dev_idx in set(self.distributed_strategy.layer_to_device):
+                self.rot_emb.compute_freqs_cis(
+                    torch.device("cuda", dev_idx), self.config.max_expected_seq_len
+                )
+        else:
+            self.rot_emb.compute_freqs_cis(
+                self.shared.emb.weight.device, self.config.max_expected_seq_len
+            )
+
+        layers = []
+        for i in range(self.config.nlayers):
+            block: nn.Module = MixtralBlock(self.config, self.rot_emb)
+            block = self.distributed_strategy.distribute_layer(block, i)
+            layers.append(block)
+        self.layers = nn.ModuleList(layers)
+
+        dec_norm = LayerNormParameterized(
+            self.config.dim,
+            elementwise_scale=True,
+            elementwise_shift=False,
+            use_mean=False,
+            eps=self.config.norm_eps,
+            use_high_precision_pow=True,
+        )
+        self.dec_norm = self.distributed_strategy.distribute_module(
+            dec_norm, final_layers=True
+        )
+
+        if self.config.p_dropout:
+            self.dropout = nn.Dropout(self.config.p_dropout)
+
+        self.reset_params()
+
+    def get_config(self) -> MixtralConfig:
+        return self.config
+
+    @classmethod
+    def from_config(cls, config: MixtralConfig) -> "Mixtral":
+        return cls(config)
+
+    def reset_params(self):
+        # Modules are self-initializing, we're just going to down-scale the final prediction head to be
+        # mixed-fan (inputs and gradients scale to the same inverse factors) if it isn't tied
+        self.shared.head.weight.data.normal_(
+            0, 1 / math.sqrt(math.sqrt(self.width * self.shared.vocab_size))
+        )
+
+    def _helper(
```

Comment:
If not too much trouble, can we separate this into MixtralHeadless and Mixtral similar to the way GPTBigCode and Roberta are implemented?